### PR TITLE
Create accumulator element (for loss) directly on the device

### LINKED ISSUE #12408: Is this a bug? - unnecessary CPU/GPU copying in supporters.py just for aggregating loss
I've noticed that while training, this chunk here is eating a lot of time.

https://github.com/PyTorchLightning/pytorch-lightning/blob/0b682b807a76b732492a3d43e3208a80c80b6c2f/pytorch_lightning/trainer/supporters.py#L76-L80

I'm wondering whether this is a bug, and why we don't have it like:

```
 self.memory = torch.zeros(self.window_length, *x.shape, device=x.device)
```
It works, and it's much, faster. 

cc @borda @akihironitta

### PR DESCRIPTION
**Pull request by isvogor-foi**
…rectly on the device

## What does this PR do?

Fixes #12408 

### Does your PR introduce any breaking changes? If yes, please list them.

No

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/supporters.py [hunk 1]
```diff
@@ -73,7 +73,7 @@ def last(self):
     def append(self, x):
         """Add an element to the accumulator."""
         if self.memory is None:
-            self.memory = torch.zeros(self.window_length, *x.shape)
+            self.memory = torch.zeros(self.window_length, *x.shape, device=x.device)
 
         # ensure same device and type
         if self.memory.device != x.device or self.memory.type() != x.type():
```
Comment (carmocca) [inline pytorch_lightning/trainer/supporters.py:79]: We can probably remove this `if` now and do

```python
        if self.memory is None:
            self.memory = torch.zeros(self.window_length, *x.shape, device=x.device)
        else:
            x = x.to(self.memory)
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/supporters.py [hunk 2]
```diff
@@ -68,16 +68,16 @@ def reset(self, window_length: Optional[int] = None) -> None:
...
-            self.memory = torch.zeros(self.window_length, *x.shape)
-
-        # ensure same device and type
-        if self.memory.device != x.device or self.memory.type() != x.type():
-            x = x.to(self.memory)
+            # tradeoff memory for speed by keeping the memory on device
+            self.memory = torch.zeros(self.window_length, *x.shape, device=x.device)
+        else:
+            # ensure same device and type
+            self.memory = self.memory.to(x)
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/supporters.py:80]: is there any reason why dtype is not specified here but in the next iteration under `else` condition?
```suggestion
            self.memory = torch.zeros(self.window_length, *x.shape, device=x.device, dtype=x.dtype)
```

**Review by rohitgr7 — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

**Comment by isvogor-foi** [comment]
@carmocca is it possible to trigger the ci-tests again? Looks like some pip dependency error was raised, which is unrelated to this PR. Build infrastructure hickup?
